### ISSUE TITLE
I want to install gradio, but fail because of the orjson dependency

### ISSUE THREAD

**Issue by robmaster2016** [mozillazg/pypy#18]
Here is the error log:

Building wheels for collected packages: matplotlib, orjson, ffmpy, contourpy, kiwisolver
  Building wheel for matplotlib (pyproject.toml) ... done
  Created wheel for matplotlib: filename=matplotlib-3.8.1-pp310-pypy310_pp73-linux_x86_64.whl size=7532562 sha256=b1cef1e1c3da5a4fe6dcb3078c2dda5a9aa3a7f56f31f0706f612e5d28c77cd8
  Stored in directory: /root/.cache/pip/wheels/e1/27/20/3654a07daf1ad83b235f52a828813cde747b13187862dafac1
  Building wheel for orjson (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for orjson (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [380 lines of output]
      Running `maturin pep517 build-wheel -i /opt/pypy3-7.3.13-1-seecker/bin/pypy3 --compatibility off`
      📦 Including license file "/tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/LICENSE-APACHE"
      📦 Including license file "/tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/LICENSE-MIT"
      🍹 Building a mixed python/rust project
      🔗 Found pyo3-ffi bindings
      🐍 Found PyPy 3.10 at /opt/pypy3-7.3.13-1-seecker/bin/pypy3
         Compiling target-lexicon v0.12.12
         Compiling once_cell v1.18.0
         Compiling serde v1.0.190
         Compiling version_check v0.9.4
         Compiling libc v0.2.149
         Compiling autocfg v1.1.0
         Compiling rustversion v1.0.14
         Compiling cfg-if v1.0.0
         Compiling itoa v1.0.9
         Compiling serde_json v1.0.107
         Compiling ryu v1.0.15
         Compiling zerocopy v0.7.15
         Compiling static_assertions v1.1.0
         Compiling smallvec v1.11.1
         Compiling encoding_rs v0.8.33
         Compiling bytecount v0.6.7
         Compiling associative-cache v2.0.0
         Compiling itoap v1.0.1
         Compiling simdutf8 v0.1.4
         Compiling ahash v0.8.6
         Compiling num-traits v0.2.17
         Compiling pyo3-build-config v0.20.0
         Compiling castaway v0.2.2
         Compiling cc v1.0.83
         Compiling chrono v0.4.31
         Compiling pyo3-ffi v0.20.0
         Compiling orjson v3.9.10 (/tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9)
         Compiling compact_str v0.7.1
         Compiling beef v0.5.2
         Compiling arrayvec v0.7.4
      error[E0425]: cannot find function `_PyDict_NewPresized` in crate `pyo3_ffi`
         --> src/deserialize/yyjson.rs:168:25
          |
      168 |         let dict = ffi!(_PyDict_NewPresized(len as isize));
          |                         ^^^^^^^^^^^^^^^^^^^ not found in `pyo3_ffi`
      
      error[E0425]: cannot find function `_PyDict_SetItem_KnownHash` in crate `pyo3_ffi`
         --> src/deserialize/yyjson.rs:178:27
          |
      178 |                 pyo3_ffi::_PyDict_SetItem_KnownHash(dict, pykey, pyval, str_hash!(pykey))
          |                           ^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `pyo3_ffi`
      
      error[E0425]: cannot find function `_PyDict_Contains_KnownHash` in crate `pyo3_ffi`
         --> src/util.rs:192:23
          |
      192 |             pyo3_ffi::_PyDict_Contains_KnownHash(
          |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `pyo3_ffi`
          |
         ::: src/serialize/per_type/dataclass.rs:60:19
          |
      60  |         } else if pydict_contains!(ob_type, SLOTS_STR) {
          |                   ------------------------------------ in this macro invocation
          |
          = note: this error originates in the macro `pydict_contains` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyDateTime_DATE_GET_TZINFO` in crate `pyo3_ffi`
         --> src/serialize/per_type/datetime.rs:173:27
          |
      173 |         let tzinfo = ffi!(PyDateTime_DATE_GET_TZINFO(self.ptr));
          |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyDateTime_DATE_GET_MINUTE`
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/datetime.rs:365:5
          |
      365 |     pub fn PyDateTime_DATE_GET_MINUTE(o: *mut PyObject) -> c_int;
          |     ------------------------------------------------------------ similarly named function `PyDateTime_DATE_GET_MINUTE` defined here
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyObject_CallMethodNoArgs` in crate `pyo3_ffi`
         --> src/util.rs:132:28
          |
      132 |           unsafe { pyo3_ffi::PyObject_CallMethodNoArgs($obj1, $obj2) }
          |                              ^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyObject_CallMethodObjArgs`
          |
         ::: src/serialize/per_type/datetime.rs:176:29
          |
      176 |               let py_offset = call_method!(self.ptr, UTCOFFSET_METHOD_STR);
          |                               -------------------------------------------- in this macro invocation
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/abstract_.rs:74:5
          |
      74  | /     pub fn PyObject_CallMethodObjArgs(
      75  | |         o: *mut PyObject,
      76  | |         method: *mut PyObject,
      77  | |         ...
      78  | |     ) -> *mut PyObject;
          | |______________________- similarly named function `PyObject_CallMethodObjArgs` defined here
          |
          = note: this error originates in the macro `call_method` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyObject_CallMethodOneArg` in crate `pyo3_ffi`
         --> src/util.rs:135:28
          |
      135 |           unsafe { pyo3_ffi::PyObject_CallMethodOneArg($obj1, $obj2, $obj3) }
          |                              ^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyObject_CallMethodObjArgs`
          |
         ::: src/serialize/per_type/datetime.rs:185:30
          |
      185 |               let method_ptr = call_method!(tzinfo, NORMALIZE_METHOD_STR, self.ptr);
          |                                ---------------------------------------------------- in this macro invocation
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/abstract_.rs:74:5
          |
      74  | /     pub fn PyObject_CallMethodObjArgs(
      75  | |         o: *mut PyObject,
      76  | |         method: *mut PyObject,
      77  | |         ...
      78  | |     ) -> *mut PyObject;
          | |______________________- similarly named function `PyObject_CallMethodObjArgs` defined here
          |
          = note: this error originates in the macro `call_method` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyObject_CallMethodNoArgs` in crate `pyo3_ffi`
         --> src/util.rs:132:28
          |
      132 |           unsafe { pyo3_ffi::PyObject_CallMethodNoArgs($obj1, $obj2) }
          |                              ^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyObject_CallMethodObjArgs`
          |
         ::: src/serialize/per_type/datetime.rs:186:29
          |
      186 |               let py_offset = call_method!(method_ptr, UTCOFFSET_METHOD_STR);
          |                               ---------------------------------------------- in this macro invocation
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/abstract_.rs:74:5
          |
      74  | /     pub fn PyObject_CallMethodObjArgs(
      75  | |         o: *mut PyObject,
      76  | |         method: *mut PyObject,
      77  | |         ...
      78  | |     ) -> *mut PyObject;
          | |______________________- similarly named function `PyObject_CallMethodObjArgs` defined here
          |
          = note: this error originates in the macro `call_method` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyObject_CallMethodOneArg` in crate `pyo3_ffi`
         --> src/util.rs:135:28
          |
      135 |           unsafe { pyo3_ffi::PyObject_CallMethodOneArg($obj1, $obj2, $obj3) }
          |                              ^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyObject_CallMethodObjArgs`
          |
         ::: src/serialize/per_type/datetime.rs:196:29
          |
      196 |               let py_offset = call_method!(tzinfo, UTCOFFSET_METHOD_STR, self.ptr);
          |                               ---------------------------------------------------- in this macro invocation
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/abstract_.rs:74:5
          |
      74  | /     pub fn PyObject_CallMethodObjArgs(
      75  | |         o: *mut PyObject,
      76  | |         method: *mut PyObject,
      77  | |         ...
      78  | |     ) -> *mut PyObject;
          | |______________________- similarly named function `PyObject_CallMethodObjArgs` defined here
          |
          = note: this error originates in the macro `call_method` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyDateTime_DATE_GET_TZINFO` in crate `pyo3_ffi`
         --> src/serialize/per_type/datetime.rs:213:31
          |
      213 |             let tzinfo = ffi!(PyDateTime_DATE_GET_TZINFO(self.ptr));
          |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyDateTime_DATE_GET_MINUTE`
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/datetime.rs:365:5
          |
      365 |     pub fn PyDateTime_DATE_GET_MINUTE(o: *mut PyObject) -> c_int;
          |     ------------------------------------------------------------ similarly named function `PyDateTime_DATE_GET_MINUTE` defined here
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyObject_CallMethodOneArg` in crate `pyo3_ffi`
         --> src/util.rs:135:28
          |
      135 |           unsafe { pyo3_ffi::PyObject_CallMethodOneArg($obj1, $obj2, $obj3) }
          |                              ^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyObject_CallMethodObjArgs`
          |
         ::: src/serialize/per_type/datetime.rs:216:33
          |
      216 |                   let py_offset = call_method!(tzinfo, UTCOFFSET_METHOD_STR, self.ptr);
          |                                   ---------------------------------------------------- in this macro invocation
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/abstract_.rs:74:5
          |
      74  | /     pub fn PyObject_CallMethodObjArgs(
      75  | |         o: *mut PyObject,
      76  | |         method: *mut PyObject,
      77  | |         ...
      78  | |     ) -> *mut PyObject;
          | |______________________- similarly named function `PyObject_CallMethodObjArgs` defined here
          |
          = note: this error originates in the macro `call_method` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyTuple_GET_ITEM` in crate `pyo3_ffi`
          --> src/serialize/per_type/numpy.rs:1144:30
           |
      1144 |         let descr_str = ffi!(PyTuple_GET_ITEM(el0, 1));
           |                              ^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyList_GET_ITEM`
           |
          ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/listobject.rs:60:5
           |
      60   |     pub fn PyList_GET_ITEM(arg1: *mut PyObject, arg2: Py_ssize_t) -> *mut PyObject;
           |     ------------------------------------------------------------------------------ similarly named function `PyList_GET_ITEM` defined here
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyTuple_GET_ITEM` in crate `pyo3_ffi`
        --> src/serialize/per_type/tuple.rs:46:33
         |
      46 |                 let elem = ffi!(PyTuple_GET_ITEM(self.ptr, i as isize));
         |                                 ^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyList_GET_ITEM`
         |
        ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/listobject.rs:60:5
         |
      60 |     pub fn PyList_GET_ITEM(arg1: *mut PyObject, arg2: Py_ssize_t) -> *mut PyObject;
         |     ------------------------------------------------------------------------------ similarly named function `PyList_GET_ITEM` defined here
      
      error[E0425]: cannot find function `_PyDict_Contains_KnownHash` in crate `pyo3_ffi`
         --> src/util.rs:192:23
          |
      192 |             pyo3_ffi::_PyDict_Contains_KnownHash(
          |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `pyo3_ffi`
          |
         ::: src/serialize/serializer.rs:123:54
          |
      123 |     if opt_disabled!(opts, PASSTHROUGH_DATACLASS) && pydict_contains!(ob_type, DATACLASS_FIELDS_STR)
          |                                                      ----------------------------------------------- in this macro invocation
          |
          = note: this error originates in the macro `pydict_contains` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0425]: cannot find function `_Py_HashBytes` in crate `pyo3_ffi`
        --> src/str/ffi.rs:20:30
         |
      20 |         let hash = pyo3_ffi::_Py_HashBytes(data_ptr, num_bytes);
         |                              ^^^^^^^^^^^^^ not found in `pyo3_ffi`
      
      error[E0609]: no field `hash` on type `pyo3_ffi::PyASCIIObject`
        --> src/deserialize/pyobject.rs:33:71
         |
      33 |     debug_assert!(unsafe { (*pykey.cast::<pyo3_ffi::PyASCIIObject>()).hash != -1 });
         |                                                                       ^^^^ unknown field
         |
         = note: available fields are: `ob_base`, `length`, `state`, `wstr`
      
      error[E0609]: no field `hash` on type `pyo3_ffi::PyASCIIObject`
         --> src/util.rs:164:59
          |
      164 |         unsafe { (*$op.cast::<pyo3_ffi::PyASCIIObject>()).hash }
          |                                                           ^^^^ unknown field
          |
         ::: src/deserialize/yyjson.rs:178:73
          |
      178 |                 pyo3_ffi::_PyDict_SetItem_KnownHash(dict, pykey, pyval, str_hash!(pykey))
          |                                                                         ---------------- in this macro invocation
          |
          = note: available fields are: `ob_base`, `length`, `state`, `wstr`
          = note: this error originates in the macro `str_hash` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0609]: no field `ob_sval` on type `PyBytesObject`
       --> src/ffi/bytes.rs:9:36
        |
      9 |     &(*op.cast::<PyBytesObject>()).ob_sval as *const c_char
        |                                    ^^^^^^^ unknown field
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyTuple_GET_ITEM` in this scope
        --> src/ffi/fragment.rs:42:24
         |
      42 |         let contents = PyTuple_GET_ITEM(args, 0);
         |                        ^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyList_GET_ITEM`
         |
        ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/listobject.rs:60:5
         |
      60 |     pub fn PyList_GET_ITEM(arg1: *mut PyObject, arg2: Py_ssize_t) -> *mut PyObject;
         |     ------------------------------------------------------------------------------ similarly named function `PyList_GET_ITEM` defined here
      
      error[E0063]: missing field `ob_pypy_link` in initializer of `pyo3_ffi::PyObject`
        --> src/ffi/fragment.rs:73:22
         |
      73 |             ob_base: PyObject {
         |                      ^^^^^^^^ missing `ob_pypy_link`
      
      error[E0609]: no field `ob_item` on type `PyListObject`
        --> src/ffi/list.rs:32:47
         |
      32 |             let elem = unsafe { *((*self.obj).ob_item).add(self.pos) };
         |                                               ^^^^^^^ unknown field
         |
         = note: available fields are: `ob_base`
      
      error[E0609]: no field `hash` on type `pyo3_ffi::PyASCIIObject`
         --> src/util.rs:195:60
          |
      195 |                 (*$obj2.cast::<pyo3_ffi::PyASCIIObject>()).hash,
          |                                                            ^^^^ unknown field
          |
         ::: src/serialize/per_type/dataclass.rs:60:19
          |
      60  |         } else if pydict_contains!(ob_type, SLOTS_STR) {
          |                   ------------------------------------ in this macro invocation
          |
          = note: available fields are: `ob_base`, `length`, `state`, `wstr`
          = note: this error originates in the macro `pydict_contains` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0609]: no field `hash` on type `pyo3_ffi::PyASCIIObject`
         --> src/util.rs:195:60
          |
      195 |                 (*$obj2.cast::<pyo3_ffi::PyASCIIObject>()).hash,
          |                                                            ^^^^ unknown field
          |
         ::: src/serialize/serializer.rs:123:54
          |
      123 |     if opt_disabled!(opts, PASSTHROUGH_DATACLASS) && pydict_contains!(ob_type, DATACLASS_FIELDS_STR)
          |                                                      ----------------------------------------------- in this macro invocation
          |
          = note: available fields are: `ob_base`, `length`, `state`, `wstr`
          = note: this error originates in the macro `pydict_contains` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      error[E0609]: no field `ob_sval` on type `PyBytesObject`
        --> src/serialize/writer.rs:45:31
         |
      45 |                 (*self.bytes).ob_sval
         |                               ^^^^^^^ unknown field
      
      error[E0609]: no field `hash` on type `pyo3_ffi::PyASCIIObject`
        --> src/str/ffi.rs:21:39
         |
      21 |         (*op.cast::<PyASCIIObject>()).hash = hash;
         |                                       ^^^^ unknown field
         |
         = note: available fields are: `ob_base`, `length`, `state`, `wstr`
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyTuple_SET_ITEM` in this scope
         --> src/lib.rs:207:9
          |
      207 |         PyTuple_SET_ITEM(args, 0, err_msg);
          |         ^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyList_SET_ITEM`
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/listobject.rs:66:5
          |
      66  |     pub fn PyList_SET_ITEM(arg1: *mut PyObject, arg2: Py_ssize_t, arg3: *mut PyObject);
          |     ---------------------------------------------------------------------------------- similarly named function `PyList_SET_ITEM` defined here
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyTuple_SET_ITEM` in this scope
         --> src/lib.rs:208:9
          |
      208 |         PyTuple_SET_ITEM(args, 1, doc);
          |         ^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyList_SET_ITEM`
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/listobject.rs:66:5
          |
      66  |     pub fn PyList_SET_ITEM(arg1: *mut PyObject, arg2: Py_ssize_t, arg3: *mut PyObject);
          |     ---------------------------------------------------------------------------------- similarly named function `PyList_SET_ITEM` defined here
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyTuple_SET_ITEM` in this scope
         --> src/lib.rs:209:9
          |
      209 |         PyTuple_SET_ITEM(args, 2, pos);
          |         ^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyList_SET_ITEM`
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/listobject.rs:66:5
          |
      66  |     pub fn PyList_SET_ITEM(arg1: *mut PyObject, arg2: Py_ssize_t, arg3: *mut PyObject);
          |     ---------------------------------------------------------------------------------- similarly named function `PyList_SET_ITEM` defined here
      
      error[E0425]: cannot find function, tuple struct or tuple variant `PyTuple_GET_ITEM` in this scope
         --> src/lib.rs:322:23
          |
      322 |             let arg = PyTuple_GET_ITEM(kwnames, i as Py_ssize_t);
          |                       ^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `PyList_GET_ITEM`
          |
         ::: /tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/include/cargo/pyo3-ffi-0.20.0/src/listobject.rs:60:5
          |
      60  |     pub fn PyList_GET_ITEM(arg1: *mut PyObject, arg2: Py_ssize_t) -> *mut PyObject;
          |     ------------------------------------------------------------------------------ similarly named function `PyList_GET_ITEM` defined here
      
      Some errors have detailed explanations: E0063, E0425, E0609.
      For more information about an error, try `rustc --explain E0063`.
      error: could not compile `orjson` due to 28 previous errors
      💥 maturin failed
        Caused by: Failed to build a native library through cargo
        Caused by: Cargo build finished with "exit status: 101": `PYO3_ENVIRONMENT_SIGNATURE="pypy-3.10-64bit" PYO3_PYTHON="/opt/pypy3-7.3.13-1-seecker/bin/pypy3" PYTHON_SYS_EXECUTABLE="/opt/pypy3-7.3.13-1-seecker/bin/pypy3" "cargo" "rustc" "--message-format" "json-render-diagnostics" "--manifest-path" "/tmp/pip-install-q2ewb9jn/orjson_dbfcb128336141c1a54042f2efe13cb9/Cargo.toml" "--release" "--lib"`
      Error: command ['maturin', 'pep517', 'build-wheel', '-i', '/opt/pypy3-7.3.13-1-seecker/bin/pypy3', '--compatibility', 'off'] returned non-zero exit status 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for orjson

Basically, thank you for your good work, pypy is really fast and good !!!
